examples/scenes/src/cosmic_text_scene.rs

```diff
@@ -0,0 +1,425 @@
+// Copyright 2024 the Vello Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use crate::SceneParams;
+
+use vello::kurbo::{Affine, Point, Rect, Size};
+use vello::peniko::{Blob, Color, Fill, Font};
+use vello::{Glyph, Scene};
+
+use cosmic_text::fontdb::ID;
+use cosmic_text::{
+    Attrs, Buffer, Cursor, Edit, Editor, FontSystem, LayoutRun, Metrics, Selection, Shaping,
+};
+
+use unicode_segmentation::UnicodeSegmentation;
+
+use std::cmp;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+pub struct CosmicTextSceneState {
+    pub font_system: FontSystem,
+    pub vello_fonts: HashMap<ID, Font>,
+    pub buffer_glyphs: BufferGlyphs,
+}
+
+impl Default for CosmicTextSceneState {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl CosmicTextSceneState {
+    pub fn new() -> Self {
+        let mut font_system = FontSystem::new();
+        let mut vello_fonts = HashMap::new();
+
+        // Copy fonts from cosmic_text, so vello can use them
+        let font_faces: Vec<(ID, u32)> = font_system
+            .db()
+            .faces()
+            .map(|face| (face.id, face.index))
+            .collect();
+
+        for (font_id, index) in font_faces {
+            if let Some(font) = font_system.get_font(font_id) {
+                let resource = Arc::new(font.data().to_vec());
```

Comment:
There are two allocations here, one to put the data in the vec, and one to allocate the `Arc`. The first could be avoided by creating an adapter which implements `AsRef<[u8]>` by containing an `Arc<Font>` and using the data parameter. The second would use https://docs.rs/bytemuck/latest/bytemuck/allocation/trait.TransparentWrapperAlloc.html#method.wrap_arc, and is definitely out-of-scope.

The best resolution to this would probably be to have a comment: "These allocations are avoidable, but used for simplicity.", maybe with a link to this comment?

Reply:
Not in time for this patch, but improving this is also a design goal for splitting these resource handles out of Peniko. (https://github.com/linebender/peniko/pull/68)

Reply:
I have used an adapter and added a link to this comment.

Reply:
A wrapper/adapter is fine, but we should probably just submit a PR to cosmic-text to make `Font` implement `AsRef<[u8]>`. Alternatively, the correct type can be obtained using `unsafe { font_system.db().make_shared_face_data(font_id).unwrap() }`.